### DISCUSSION TITLE
[A-L] (30 maggio 2024) Esercizio 6

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/progettazione-di-algoritmi#109] [Esercitazioni A-L]
![immagine](https://github.com/user-attachments/assets/56785581-99b4-4836-aec0-884e667a0952)

**Comment by Elia-Belli** [comment]
Il numero $n$ può essere ottenuto in 3 modi applicando le rispettive operazioni ai numeri $n-1,\ n/2,\ n/3$ , ottenibili ricorsivamente allo stesso modo.
- Se vogliamo minimizzare le operazioni sceglieremo ad ogni chiamata ricorsiva il minimo dei 3 per poi sommare $1$, ovvero l'ultima operazione necessaria per ottenere $n$, con caso base $x_2 = 0$
$$x_n = min(x_{n-1},\ x_{n/2},\ x_{n/3}) +1$$

- Se vogliamo calcolare il numero di modi per ottenere $n$ dovremo sommare i modi per ottenere i numeri a cui applicare l'ultima operazione ricorsivamente, con caso base $y_2 = 1$
$$y_n = y_{n-1} + y_{n/2} + y_{n/3}$$

**Comment by CiottoloMaggico** [comment]
![immagine](https://github.com/user-attachments/assets/d2d9d5f9-5696-4381-b80d-1ddc6b3d5893)

**Comment by daromacarlo** [comment]
è possibile trovare Xn con un algoritmo di programmazione dinamica
![immagine](https://github.com/user-attachments/assets/c53eb215-3035-4608-a26f-14fab7f6e86a)
ma non è possibile trovare Yn con un algoritmo di programmazione dinamica nel tempo richiesto, ho quindi usato una strategia differente

![immagine](https://github.com/user-attachments/assets/b2519ab2-ea1f-4a73-a8ee-1f09ef5b7ca2)
![immagine](https://github.com/user-attachments/assets/e5486ca8-b42b-4125-b928-a78127815d57)
queste funzioni hanno costo O(n + 3n)

![immagine](https://github.com/user-attachments/assets/7edc0947-f27d-413d-9868-c8c799039f5a)
